ntp-proto/src/timetypes.rs

```diff
@@ -16,10 +21,48 @@ impl NtpTimestamp {
         self.timestamp.to_be_bytes()
     }
 
+    pub(crate) fn from_system_time(time: std::time::SystemTime) -> Self {
+        let dur = time.duration_since(std::time::UNIX_EPOCH).unwrap();
+        let secs = dur.as_secs() + EPOCH_OFFSET;
+        let nanos = dur.subsec_nanos();
+
+        Self::from_seconds_nanos_since_ntp_epoch(secs, nanos)
+    }
+
+    pub(crate) fn from_seconds_nanos_since_ntp_epoch(seconds: u64, nanos: u32) -> Self {
+        // NTP uses 1/2^32 sec as its unit of fractional time.
+        // our time is in nanoseconds, so 1/1e9 seconds
+        let fraction = ((nanos as u64) << 32) / 1_000_000_000;
+
+        // alternatively, abuse FP arithmetic to save an instruction
+        // let fraction = (nanos as f64 * 4.294967296) as u64;
+
+        let timestamp = (seconds << 32) + fraction;
+        NtpTimestamp::from_bits(timestamp.to_be_bytes())
+    }
+
+    #[cfg(test)]
+    pub(crate) fn duration_since_unix_epoch(self) -> std::time::Duration {
```

Comment:
I feel like there's already too much code here that is `#[cfg(test)] pub(crate)`. It can sometimes be found, but most of the code that is only necessary for tests should live in one of the tests modules, not in the library. If need be, just make the `NtpTimestamp::timestamp` field `pub(crate)` (though it might be better to put tests that need access like that in a test module inside this module, which can access the field even if private).